ldm/dream/args.py

```diff
@@ -0,0 +1,485 @@
+
+    def parse_args(self):
+        '''Parse the shell switches and store.'''
+        self._arg_switches = self._arg_parser.parse_args()
+        return self._arg_switches
+
+    def parse_cmd(self,cmd_string):
+        '''Parse a dream>-style command string '''
+        command = cmd_string.replace("'", "\\'")
+        try:
+            elements = shlex.split(command)
+        except ValueError:
+            print(traceback.format_exc(), file=sys.stderr)
+            return
+        switches = ['']
+        switches_started = False
+
+        for el in elements:
+            if el[0] == '-' and not switches_started:
+                switches_started = True
+            if switches_started:
+                switches.append(el)
+            else:
+                switches[0] += el
+                switches[0] += ' '
+        switches[0] = switches[0][: len(switches[0]) - 1]
+        self._cmd_switches = self._cmd_parser.parse_args(switches)
+        return self._cmd_switches
+
+    def json(self,**kwargs):
+        return json.dumps(self.to_dict(**kwargs))
+
+    def to_dict(self,**kwargs):
+        a = vars(self)
+        a.update(kwargs)
+        return a
+
+    # Isn't there a more automated way of doing this?
+    # Ideally we get the switch strings out of the argparse objects,
+    # but I don't see a documented API for this.
+    def prompt_str(self,**kwargs):
+        """Normalized prompt."""
+        a = vars(self)
+        a.update(kwargs)
+        switches = list()
+        switches.append(f'"{a["prompt"]}')
+        switches.append(f'-s {a["steps"]}')
+        switches.append(f'-W {a["width"]}')
+        switches.append(f'-H {a["height"]}')
+        switches.append(f'-C {a["cfg_scale"]}')
+        switches.append(f'-A {a["sampler_name"]}')
+        switches.append(f'-S {a["seed"]}')
+        if a['seamless']:
+            switches.append(f'--seamless')
+        if len(a['init_img'])>0:
+            switches.append(f'-I {a.init_img}')
+        if a['fit']:
+            switches.append(f'--fit')
+        if a['strength'] and len(a['init_img'])>0:
+            switches.append(f'-f {a["strength"]}')
+        if a['gfpgan_strength']:
+            switches.append(f'-G {a["gfpgan_strength"]}')
+        if a['upscale']:
+            switches.append(f'-U {" ".join([str(u) for u in a["upscale"]])}')
+        if a['embiggen']:
+            switches.append(f'--embiggen {" ".join([str(u) for u in a["embiggen"]])}')
+        if a['embiggen_tiles']:
+            switches.append(f'--embiggen_tiles {" ".join([str(u) for u in a["embiggen_tiles"]])}')
+        if a['variation_amount'] > 0:
+            switches.append(f'-v {a["variation_amount"]}')
+        if a['with_variations']:
+            formatted_variations = ','.join(f'{seed}:{weight}' for seed, weight in (a["with_variations"]))
+            switches.append(f'-V {formatted_variations}')
+        return ' '.join(switches)
+
+
+    def __getattribute__(self,name):
```

Comment:
Python has easier ways of doing getters and setters, but it's no big deal.

Reply:
Do tell! I'm learning.

Reply:
There's a bit of explaining prior to [this](https://realpython.com/python-property/#using-property-as-a-decorator) discussion (explaining the theory behind Pythonic properties), but I think you're advanced enough not to need it, and you can probably even scroll past the first few paragraphs to the Circle source listing (unless you've never used decorators).

Reply:
Python is such a neat programming language. So many interesting avenues to explore.

I don't think that the getter setter decorators quite work for what I'm trying to do, which is to create an object that is a union of the shell command-line arguments and the dream command line and has rules for selecting among them, but I'm keen to try them in other contexts.

BTW, my original __getattribute__ code caused a horrendous infinite recursion when trying to execute a copy.copy() on the object. I looked at the copy.copy() code and there is (to me) some deep voodoo going on there. I had to write a bit of defensive code in order to prevent the half-constructed copied object from blowing up.

Reply:
> I don't think that the getter setter decorators quite work for what I'm trying to do

I had a feeling that might be the case, but I wanted you to make the call :)

> my original getattribute code caused a horrendous infinite recursion when trying to execute

I might be thinking of the wrong thing generally here, but would `deepcopy` help?